### ISSUE TITLE
Automated feature detection (and matching) - long-term

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#300]
Let's not think too much about this YET - but one add-on feature we could add to this feature for image distortion (#299) would be a way to automatically examine 2 images and find how they ought to overlap. This has several steps:

1. finding interest points in 2 images
2. matching interest points between 2 images
3. calculating the corner transformations required to use the matches (if any) to perform a distortion using #299

Here's one in-depth tutorial that goes through all the necessary steps: https://hub.packtpub.com/object-detection-using-image-features-javascript/

![image](https://user-images.githubusercontent.com/24359/41996960-834ca58e-7a24-11e8-8d4e-e79625254997.png)

This would be great to develop as an independent module, outside of Image Sequencer!

**Comment by jywarren** [comment]
This should be possible in OpenCV.js!

* http://answers.opencv.org/question/5348/evaluation-of-interest-point-detectors-and-descriptors/
* https://docs.opencv.org/2.4/doc/tutorials/features2d/feature_detection/feature_detection.html
* https://stackoverflow.com/questions/23637916/find-interest-point-in-surf-detector-algorithm
* https://stackoverflow.com/questions/29133085/what-are-keypoints-in-image-processing

steps might be:

1. run SURF on Image 1 to find interest points
2. run SURF on Image 2 to find interest points
3. find best matches of common points from each
4. in relation to image 1, rotate, scale, and translate image 2 to best overlap the top 2-3 common points between the 2 images
5. flatten into 1 image

Amir shared: https://docs.opencv.org/3.3.0/dc/dc3/tutorial_py_matcher.html

https://docs.opencv.org/2.4/modules/ocl/doc/feature_detection_and_description.html?highlight=feature%20description

**Comment by jywarren** [comment]
An initial version wouldn't need to do full 4-corner transform but could start with just rotation/scale/translate!

**Comment by rockstar777** [comment]
Can I work on this issue?

**Comment by jywarren** [comment]
That'd be really awesome. Have you been able to merge a more basic module first? Perhaps #165 has some, if you haven't (apologies, it's been a busy week so I forget!)

However, we'd really like your help with this. It seems like a big project but we'd love to hear how you might try to solve it? What would be your first step?

👍 🎉

**Comment by rockstar777** [comment]
Thanks
I had'nt merge that before but i would try that in #165 
Yes, It is big project but let's try to solve by detecting interesting points in two images using opencv.js.
My first for this issue would be I will get comfortable in OpenCV.js, then try to implement code, then moving to matching points and distortion of images.

**Comment by jywarren** [comment]
yes, i think we could have a "finding points" module, a "matching points"
module, and then a distortion one... solve it piece by piece in separate
PRs, you know?

**Comment by jywarren** [comment]
There are actually feature detection and description abilities in `jsfeat` here: https://github.com/publiclab/image-sequencer/issues/165

I /believe/ this is in pure javascript!

**Comment by jywarren** [comment]
and there's a very cool demo that seems to be running here: https://inspirit.github.io/jsfeat/sample_orb.html

**Comment by rockstar777** [comment]
hi @jywarren,
I have created a GSOC proposal for this issue, please review and help me to improve my proposal.
Thanks (apologies for being inactive for last few months)
:)

**Comment by rockstar777** [comment]
Also, please mention if it is okay to include this issue in GSOC '19.
Thanks

**Comment by jywarren** [comment]
Of course it's fine to include! Did you post your proposal on our website
at
https://publiclab.org/notes/warren/02-28-2019/call-for-summer-of-code-2019-proposals
? Thank you!

**Comment by rockstar777** [comment]
Yes,I had Posted it from username puneetbindal8989.
Link:  [post link](https://publiclab.org/notes/puneetbindal8989/04-01-2019/gsoc-proposal-automated-feature-detection-and-matching-300-https-github-com-publiclab-image-sequencer-issues-300)
Thanks :)
